[Ng Jun Kang] iP

ngjunkang opened the pull request:
## Baron ChatBot (BCB)
> "I am a goal setter and I set more goals everyday. I keep lists of goals in my office to stay on track." - Kiana Tom ([source](https://quotefancy.com/quote/1445196/Kiana-Tom-I-am-a-goal-setter-and-I-set-more-goals-everyday-I-keep-lists-of-goals-in-my))

**BCB** allows fellow programmers like us to practise the habit of keeping "list of goals" to "stay on track". When programming, we often lose track of our goals and the time, and we forget what we were trying to achieve at the start. BCB is:
- Perfect for programmers
- Similar to our working environment (CLI)
- Fast with hardly any downtime

#### All you need to do is:
1. Download the jar file from the release page [here](https://github.com/ngjunkang/ip/releases).
2. Run it with the `java -jar` command, e.g. `java -jar baron.jar`.
3. Add your tasks.
4. Let BCB manage all your tasks.

And most importantly, unlike the applications in the market, it is 💯% **FREE**!!

#### Features:
- [x] Add ToDo
- [x] Add Event
- [x] Add Deadline
- [x] List all tasks
- [x] Mark task
- [x] Un-mark task
- [x] Delete task

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    new Baron("data/baron.txt").start();
}
```

Commit:
Add Gradle support

ngjunkang committed:
Changed file structure, by moving Duke into a package

ngjunkang committed:
Duke Level-1: Greet, Echo, Exit

ngjunkang committed:
Implemented Duke Level-2: Add, List

ngjunkang committed:
Implemented Duke Level-3: Mark as Done

ngjunkang committed:
Added TaskType enum, consisting of TODO, DEADLINE, EVENT

ngjunkang committed:
Implement Duke Level-4: ToDo, Event, Deadline

ngjunkang committed:
Implement Duke Level-5: Handle Errors

ngjunkang committed:
Implement Duke Level-6: Delete

ngjunkang committed:
Add tests case and expected output for automated Text UI Testing

ngjunkang committed:
Add exceptions for deadline and event, eg deadline /by date, deadline desc /by

ngjunkang committed:
Update AddTaskCommand to add into storage file

ngjunkang committed:
Update DeleteTaskCommand to update the storage file as well

ngjunkang committed:
Add StorageManager dependency to CommandManager to enable file storage

ngjunkang committed:
Add toSaveString function to output string to be stored

ngjunkang committed:
Add the capability to update the data file during the command execute()

ngjunkang committed:
Add getAllTask(), TaskManager(ArrayList<Task>), revertChanges() and commitChanges for storage operations

ngjunkang committed:
Add storage/saving functionality to Baron

ngjunkang committed:
Add error messages of the storage/saving functionality

ngjunkang committed:
Update the by private variable to be of LocalDateTime type and change toString()

ngjunkang committed:
Update the at private variable to be of LocalDateTime type and change toString()

ngjunkang committed:
Add error message for invalid date/time format

ngjunkang committed:
Update addTask() to handle Date/Time parsing to pass to Event and Deadline

ngjunkang committed:
Add utility to handle LocalDateTime objects

ngjunkang committed:
Merge branch 'branch-Level-7'

ngjunkang committed:
Merge branch 'branch-Level-8'

ngjunkang committed:
Combine the functionality of Level-7 and Level-8

ngjunkang committed:
Set previousTaskList to null at the end of revertChanges()

ngjunkang committed:
Refactor StorageManager to Storage

ngjunkang committed:
Refactor Printer class to TextUi class

ngjunkang committed:
Rename package messages.Messages to message.Message

ngjunkang committed:
Add JUnit Tests for Todo and DateTimeUtil

ngjunkang committed:
Add JavaDoc comments to AddTaskCommand

ngjunkang committed:
Add JavaDoc comments to Baron

ngjunkang committed:
Add JavaDoc comments to BaronException

ngjunkang committed:
Add JavaDoc comments to ByeCommand

ngjunkang committed:
Add JavaDoc comments to Command

ngjunkang committed:
Add JavaDoc comments to CommandManager

ngjunkang committed:
Add JavaDoc comments to CommandParser

ngjunkang committed:
Add JavaDoc comments to DateTimeUtil

ngjunkang committed:
Add JavaDoc comments to Deadline

ngjunkang committed:
Add JavaDoc comments to DeleteTaskCommand

ngjunkang committed:
Add JavaDoc comments to EchoCommand

ngjunkang committed:
Add JavaDoc comments to EmptyCommand

ngjunkang committed:
Add JavaDoc comments to Event

ngjunkang committed:
Add JavaDoc comments to InvalidCommand

ngjunkang committed:
Add JavaDoc comments to ListCommand

ngjunkang committed:
Add JavaDoc comments to MarkTaskCommand

ngjunkang committed:
Add JavaDoc comments to Message

ngjunkang committed:
Add JavaDoc comments to Storage

ngjunkang committed:
Add JavaDoc comments to Task

ngjunkang committed:
Add JavaDoc comments to TaskManager

ngjunkang committed:
Add JavaDoc comments to TextUi

ngjunkang committed:
Add JavaDoc comments to ToDo

ngjunkang committed:
Add JavaDoc comments to UnmarkTaskCommand

ngjunkang committed:
Make minor edits according to coding standard

ngjunkang committed:
Add FindCommand to commands package

ngjunkang committed:
Add FindCommand to parseCommand method

ngjunkang committed:
Add filterDescByKeyword method with Javadocs

ngjunkang committed:
Merge branch 'branch-A-JavaDoc'

ngjunkang committed:
Merge branch 'branch-A-CodingStandard'

ngjunkang committed:
Merge branch 'branch-Level-9'

tandeshao commented on src/main/java/baron/util/TextUi.java:
```diff
@@ -0,0 +1,29 @@
+import baron.messages.Messages;
+import baron.tasks.TaskManager;
+
+public class TextUi {
```
I like how you have abstracted out the UI from Duke main logic 😄 !

ngjunkang replied:
Thank you! :)

tandeshao commented on src/main/java/baron/messages/Messages.java:
```diff
@@ -8,15 +8,20 @@
+    public static final String MESSAGE_INVALID_FILE_FORMAT = "Data file format is invalid.";
+    public static final String MESSAGE_FILE_CREATION_FAIL = "Data file cannot be created.";
+    public static final String MESSAGE_FILE_WRITE_FAIL = "Data file cannot be written to.";
     public static final String MESSAGE_DATE_TIME_FORMAT_INVALID = "The date/time should be in the following format: d/M/yyyy HH:mm";
```
I didn't know you could name all the default messages as an attribute, I'm impressed!

ngjunkang replied:
Hi De Shao, my reason for this design is actually just for me to easily change the outputs, instead of finding different Java files to fix the outputs. For example, if I want to change to a more kids-friendly tone. Glad that you are impressed!

tandeshao commented on src/main/java/baron/commands/CommandManager.java:
```diff
@@ -62,6 +62,8 @@ public Command parseCommand(String fullCommand) {
                 return new DeleteTaskCommand(this.taskManager, this.storage, commandArg);
+            case "find":
+                return new FindCommand(this.taskManager, commandArg);
             default:
```
In Intellij they allow enhanced switch-case statements, you might want to change your switch-case control flow for better readability. An example of enhanced switch-case statements:

```java
switch (itemCode) {
    case 001 -> System.out.println("It's a laptop!");
    case 002 -> System.out.println("It's a desktop!");
    case 003,004 -> System.out.println("It's a mobile phone!");
}
```

ngjunkang replied:
Hi De Shao, I didn't know there's such a syntax in Java, looks similar to lambdas. Good to know! And I will definitely use in the future. However, I won't be applying to my code, because I am not sure it is compliant to the coding standard and the Java version (JDK 11, it's a JDK 12-13 feature, I think) of this course. Thank you for the suggestion though.

tandeshao commented on src/main/java/baron/tasks/Task.java:
```diff
@@ -29,10 +54,21 @@ public boolean unmark() {
+     * @param delimiter the string to be inserted in between various information of this task.
+     * @return the encoded string to be stored in the storage using the specified delimiter.
+     */
     public String toSaveString(String delimiter) {
```
I believe ``saveString`` is sufficient.

ngjunkang replied:
Hi De Shao, thank you for the suggestion, but I feel that as this method doesn't save string, so `saveString` would be misleading (it may mean that this method saves the string somewhere, `save` as in the verb). In fact, it just returns the encoded string that is to be saved somewhere by other parts of my code.

tandeshao commented on src/main/java/baron/commands/UnmarkTaskCommand.java:
```diff
@@ -5,22 +5,41 @@
     private final TaskManager taskManager;
     private final Storage storage;
-    private final String commandArgs;
+    private final String commandArg;
```
I like how you change the name of ``commandArgs`` to ``commandArg`` since ``commandArg`` only represents a single command

ngjunkang replied:
Thank you! :)

tandeshao commented on src/main/java/baron/commands/CommandManager.java:
```diff
@@ -33,20 +33,22 @@ public Command parseCommand(String fullCommand) {
-                    return new DeleteTaskCommand(this.taskManager, this.storage, commandArg);
-                default:
-                    return new InvalidCommand();
+            case "mark":
+                return new MarkTaskCommand(this.taskManager, this.storage, commandArg);
+            case "unmark":
+                return new UnmarkTaskCommand(this.taskManager, this.storage, commandArg);
+            case "todo":
+                return new AddTaskCommand(this.taskManager, this.storage, TaskType.TODO, commandArg);
+            case "deadline":
+                return new AddTaskCommand(this.taskManager, this.storage, TaskType.DEADLINE, commandArg);
+            case "event":
+                return new AddTaskCommand(this.taskManager, this.storage, TaskType.EVENT, commandArg);
+            case "delete":
+                return new DeleteTaskCommand(this.taskManager, this.storage, commandArg);
+            case "find":
+                return new FindCommand(this.taskManager, commandArg);
+            default:
+                return new InvalidCommand();
```
Consider using enhanced switch-case statements.

tandeshao reviewed (commented):
Code looks clean and succinct 😄 ! LGTM

ngjunkang committed:
Merge branch 'add-gradle-support'

ngjunkang committed:
Change Gradle build configuration to baron.Baron

ngjunkang committed:
Add CheckStyle configurations

ngjunkang committed:
Fix coding style violations using CheckStyle

lawwm commented on src/main/java/baron/tasks/TaskManager.java:
```diff
@@ -0,0 +1,271 @@
+     *
+     * @param keyword the keyword to filter the task list by.
+     * @return the filtered task list.
+     */
```
Good use of fp.

lawwm commented on src/main/java/baron/tasks/TaskManager.java:
```diff
@@ -0,0 +1,271 @@
+     * @see #taskList
+     * @see #previousTaskList
+     * @see #commitChanges()
+     */
```
Seems like it depends on client using it correctly (after an edit operation). Maybe can return boolean value depending on if there is a previous change to revert to?

lawwm commented on src/main/java/baron/tasks/TaskManager.java:
```diff
@@ -0,0 +1,271 @@
+     * @return true if the task is un-marked successfully.
+     * @throws BaronException If there is no task, the specified task was not marked,
+     *                        or the index is out of bound.
+     */
```
Maybe you can use guard clauses to make happy path more prominent?

lawwm reviewed (commented):
~~Removed.~~

lawwm commented on src/main/java/baron/commands/DeleteTaskCommand.java:
```diff
@@ -0,0 +1,57 @@
+     *
+     * @return the delete task command output.
+     */
+    @Override
+    public String execute() {
+        try {
+            int index = CommandParser.parseTaskIntArg(this.commandArg);
+            Task deletedTask = this.taskManager.deleteTask(index);
+            try {
+                this.storage.save(this.taskManager.getAllTasks());
+            } catch (BaronException e) {
+                this.taskManager.revertChanges();
+                throw e;
+            }
+            this.taskManager.commitChanges();
+            return Message.MESSAGE_DELETE_SUCCESS + deletedTask.toString() + "\n"
+                    + Message.generateNoOfTasksMessage(this.taskManager.getTaskCount());
+        } catch (BaronException e) {
+            return e.toString();
+        }
+    }
```
Is it possible to refactor the nested try catch statements into one try catch statement for better readability?

lawwm commented on src/main/java/baron/tasks/TaskManager.java:
```diff
@@ -0,0 +1,271 @@
+     */
+    public ArrayList<Task> filterDescByKeyword(String keyword) {
+        ArrayList<Task> filteredTaskList = new ArrayList<>();
+        this.taskList.forEach((task -> {
+            if (task.description.contains(keyword)) {
+                filteredTaskList.add(task);
+            }
+        }));
```
Good use of fp.

lawwm commented on src/main/java/baron/tasks/TaskManager.java:
```diff
@@ -0,0 +1,271 @@
+     * @see #previousTaskList
+     * @see #commitChanges()
+     */
+    public void revertChanges() {
+        if (this.previousTaskList != null) {
+            this.taskList = new ArrayList<>(this.previousTaskList);
+            this.previousTaskList = null;
+        }
+    }
```
Seems like it depends on client using it correctly (after an edit operation). Maybe can return boolean value depending on if there is a previous change to revert to?

lawwm commented on src/main/java/baron/tasks/TaskManager.java:
```diff
@@ -0,0 +1,271 @@
+     * @throws BaronException If there is no task, the specified task was already marked,
+     *                        or the index is out of bound.
+     */
+    public boolean markTask(int index) throws BaronException {
+        if (this.isEmpty()) {
+            throw new BaronException(Message.MESSAGE_NO_TASK);
+        } else if (index > 0 && index <= this.getTaskCount()) {
+            if (this.taskList.get(index - 1).mark()) {
+                return true;
+            } else {
+                throw new BaronException("The task is already marked.");
+            }
+        } else {
+            throw new BaronException(Message.generateTaskIndexOutOfBoundMessage(this.getTaskCount()));
+        }
+    }
```
Maybe you can use guard clauses to make happy path more prominent? This can be applied for other methods in this class as well.

ngjunkang replied:
Hi, can you provide me with an example of a guard clause? :)

lawwm reviewed (commented):
Looks good, didn't notice coding standard violation. Sorry for reposting, I made a mistake on first review.

ngjunkang committed:
Add JavaFX tutorial part 1

ngjunkang committed:
Add JavaFX tutorial part 2

Jacky142857 reviewed (commented):
This is almost impeccable! Very good use of constants and JavaDoc you wrote is the best I've seen. You used {@code } to highlight the keyword in javadoc and @see to guide us to other descriptions. I think some nested if statement and try statement can be written in parallel to increase readability. You did quite a good job.

ngjunkang committed:
Add JavaFX tutorial part 4

ngjunkang committed:
Update JavaFX tutorial part 4

ngjunkang committed:
Change application to GUI

ngjunkang committed:
Fix coding style violations

ngjunkang committed:
Remove emoji for error message

ngjunkang committed:
Improve the overall appearance of GUI

ngjunkang committed:
Merge branch 'branch-Level-10'

ngjunkang committed:
Fix JavaDoc method header format

ngjunkang committed:
Change test cases due to edits in BaronException

ngjunkang committed:
Add assertions

Some assumptions are were in the code, so by adding assertions, we can
verify them at runtime and handle the error if necessary.

Let's,
* add assertion to parseCommand in CommandManager
* add assertion to generateTaskIndexOutOfBoundMessage() in Message
* add assertions to getDisplayString(), getSaveString() in DateTimeUtil
* handle null in getDisplayString(), getSaveString() in DateTimeUtil
* add assertion to generateTaskIndexOutOfBoundMessage() in Message
* add assertion to addTask method in TaskManager due to null
* add assertion to markTask() in TaskManager due to getTaskCount()
* add assertion to unmarkTask() in TaskManager due to getTaskCount()
* add assertion to deleteTask(Task) in TaskManager due to getTaskCount()
* add assertion to revertChanges() in TaskManager due to null
* add assertion to commitChanges() in TaskManager due to null

This assertion update mainly deals with null and getTaskCount() in
TaskManager.

ngjunkang committed:
Improve code quality

ngjunkang committed:
Merge pull request #1 from ngjunkang/branch-A-CodeQuality

Improve code quality

ngjunkang committed:
Merge branch 'master' into branch-A-Assertions

ngjunkang committed:
Merge pull request #2 from ngjunkang/branch-A-Assertions

Add assertions

ngjunkang committed:
Add the ability to detect task duplicates

Duplicated task (Deadline, ToDo, Event) can be added into the
TaskManager, which is redundant to the user.

This update prevents user from adding duplicated tasks.

Let's,
* add an abstract equals(Object) method to Task
* implement equals(Object) method in Task's subclasses
* add a duplicate check in addTask(TaskType, String) in TaskManager
* add generateDuplicateTaskMessage(TaskType) in Message

This uses the power of polymorphism by leveraging the use of the equal
method of Object and overriding it in subclasses, so minimal edits were
made.

Event objects are equal when description and at variables are equal.

Deadline objects are equal when description and at variables are equal.

ToDo objects are equal when description variables are equal

ngjunkang committed:
Merge pull request #3 from ngjunkang/branch-C-DetectDuplicates

Add the ability to deal with task duplicates

ngjunkang committed:
Set up Continuous Integration (CI)

ngjunkang committed:
Fix long methods and JavaDoc description

ngjunkang committed:
Add assertions to Storage

ngjunkang committed:
Add brackets to conditions

ngjunkang committed:
Add test cases to replace text ui test

ngjunkang committed:
Separate test cases into methods

ngjunkang committed:
Improve the GUI

ngjunkang committed:
Fix test cases to match GUI

ngjunkang committed:
Add a representative screenshot

ngjunkang committed:
Fix index and case sensitivity of find command

ngjunkang committed:
Add a brief User Guide (UG)

ngjunkang committed:
Set theme jekyll-theme-hacker

ngjunkang committed:
Update User Guide

ngjunkang committed:
Merge branch 'master' of https://github.com/ngjunkang/ip

ngjunkang committed:
Add User Guide header

ngjunkang committed:
Remove User Guide header

ngjunkang committed:
Set theme jekyll-theme-architect

ngjunkang committed:
Merge branch 'master' of https://github.com/ngjunkang/ip

ngjunkang committed:
Remove User Guide title

ngjunkang committed:
Add space after index when listing

ngjunkang committed:
Update Task index out of bound message

ngjunkang committed:
Add command summary
